File: 03-TP/donnees.py
"""Chargement des séries du cours — corrigés Python.

Économétrie des séries temporelles — M1 MBFA

Les chemins sont relatifs à la racine du dépôt. Si un fichier local manque,
on bascule sur l'URL GitHub : les corrigés fonctionnent donc sur Google Colab
sans clone préalable.
"""

from pathlib import Path
import pandas as pd

BASE_URL = ("https://raw.githubusercontent.com/bilelsanhaji/"
            "Cours-Series-Temporelles/refs/heads/main/04-Data/")


def _source(fichier: str, sous_dossier: str = "") -> str:
    local = Path("04-Data") / sous_dossier / fichier
    if local.exists():
        return str(local)
    suffixe = f"{sous_dossier}/" if sous_dossier else ""
    return BASE_URL + suffixe + fichier


def lire_insolation(fichier: str = "SH_MIN006088001.csv",
                    sous_dossier: str = "") -> pd.DataFrame:
    """Insolation mensuelle Météo-France. Renvoie les colonnes date / valeur."""
    d = pd.read_csv(_source(fichier, sous_dossier), sep=";", comment="#")
    d = d.dropna(subset=["VALEUR"])
    return pd.DataFrame({
        "date":   pd.to_datetime(d["YYYYMM"].astype(int).astype(str), format="%Y%m"),
        "valeur": pd.to_numeric(d["VALEUR"]),
    }).reset_index(drop=True)


def annuel(d: pd.DataFrame) -> pd.DataFrame:
    """Cumul annuel : somme des douze mois. 84 observations, 1931-2014."""
    g = d.assign(annee=d["date"].dt.year).groupby("annee")["valeur"]
    complet = g.count() == 12               # on écarte les années tronquées
    return (g.sum()[complet].rename("valeur").reset_index())


def lire_externe(nom: str) -> pd.DataFrame:
    """Séries FRED / Yahoo mises en cache par
    04-Data/mettre-a-jour-donnees-externes.R"""
    f = Path("04-Data") / f"{nom}.csv"
    if not f.exists():
        raise FileNotFoundError(
            f"Cache absent : {f}\n  Lancez une fois, depuis la racine du dépôt :\n"
            '    source("04-Data/mettre-a-jour-donnees-externes.R")')
    d = pd.read_csv(f)
    d["date"] = pd.to_datetime(d["date"])
    return d

File: 03-TP/donnees.R
# Chargement des séries du cours — corrigés R
# Économétrie des séries temporelles — M1 MBFA
#
# Les chemins sont relatifs à la racine du dépôt (execute-dir: project).
# Si un fichier local manque, on bascule automatiquement sur l'URL GitHub :
# les corrigés fonctionnent donc aussi sur Posit Cloud sans clone.

BASE_URL <- paste0("https://raw.githubusercontent.com/bilelsanhaji/",
                   "Cours-Series-Temporelles/refs/heads/main/04-Data/")

.source_donnees <- function(fichier, sous_dossier = "") {
  local <- file.path("04-Data", sous_dossier, fichier)
  if (file.exists(local)) local else paste0(BASE_URL, sous_dossier,
                                            if (nzchar(sous_dossier)) "/" else "", fichier)
}

# Insolation mensuelle Météo-France : renvoie date / valeur
lire_insolation <- function(fichier = "SH_MIN006088001.csv", sous_dossier = "") {
  d <- read.table(.source_donnees(fichier, sous_dossier), sep = ";", header = TRUE,
                  comment.char = "#", stringsAsFactors = FALSE)
  d <- d[!is.na(suppressWarnings(as.numeric(d$VALEUR))), ]
  data.frame(date   = as.Date(paste0(d$YYYYMM, "01"), format = "%Y%m%d"),
             valeur = as.numeric(d$VALEUR))
}

# Cumul annuel : somme des douze mois. 84 observations, 1931-2014.
annuel <- function(d) {
  an <- as.integer(format(d$date, "%Y"))
  n  <- tapply(d$valeur, an, length)
  a  <- sort(unique(an))
  complet <- a[n[as.character(a)] == 12]          # on écarte les années tronquées
  data.frame(annee  = complet,
             valeur = as.numeric(tapply(d$valeur, an, sum)[as.character(complet)]))
}

# Séries externes mises en cache (FRED, Yahoo). Voir
# 04-Data/mettre-a-jour-donnees-externes.R
lire_externe <- function(nom) {
  f <- file.path("04-Data", paste0(nom, ".csv"))
  if (!file.exists(f))
    stop("Cache absent : ", f, "\n  Lancez une fois :\n",
         '    source("04-Data/mettre-a-jour-donnees-externes.R")')
  d <- read.csv(f, stringsAsFactors = FALSE)
  d$date <- as.Date(d$date)
  d
}
